### ISSUE TITLE
`$fetch`: use `useRequestFetch()` when calling internal API during SSR

### ISSUE THREAD

**Issue by Atinux** [nuxt/nuxt#24813]
### Describe the feature

When having API routes in Nuxt with authentication (using H3 `useSession` or [nuxt-auth-utils](https://github.com/atinux/nuxt-auth-utils)), we need to forward the `cookie` to the API routes when fetching the data on SSR.

### Not working

```vue
<script setup>
const { data: teams } = await useAsyncData('team', () => $fetch('/api/teams')
</script>
```

### Working

```vue
<script setup>
const { data: teams } = await useAsyncData('team', () => useRequestFetch()('/api/teams')
</script>
```

Or using `useFetch`:

```vue
<script setup>
const { data: teams } = await useFetch('/api/teams', { key: 'teams' })
</script>
```

I believe it should work when we call our Nuxt API directly with `$fetch` to improve the DX.

### Additional information

- [X] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by pi0** [comment]
This is a nice idea to improve DX, considering cases where users are directly using `$fetch` for this purpose of passing header however I'm afraid it reduces the portability of `fetch` and `$fetch` if we do an implicit behavior like this (ie: a utility can work in ambient context and plugins while broken in lifecycle of a request with invalid headers or wise-versa)

I would suggest alternatively this to be shipped as part of factory function fetches. We could introduce also a preconfigured `$apiFetch` version that is intended for end-users explicitly and can contain additional logic like this.

**Comment by danielroe** [comment]
We agreed to defer discussion on this until https://github.com/nuxt/nuxt/issues/14736.

**Comment by blqke** [comment]
Had this problem using lucia auth example, with this server middleware: https://github.com/lucia-auth/examples/blob/main/nuxt/github-oauth/server/middleware/auth.ts

when using `$fetch` inside `useAsyncData`,
event.context ⤵️
```
{
  _nitro: { routeRules: {} },
  nitro: { errors: [] },
  session: {
    id: 'something',
    userId: 'something',
    fresh: false,
    expiresAt: 2024-03-23T16:22:58.000Z
  },
  user: { githubId: undefined, username: 'something', id: 'something' },
  matchedRoute: { path: '/api/any-api-route', handlers: { get: [Function] } },
  params: {}
}
```

when using `useRequestFetch()` inside `useAsyncData`,
event.context ⤵️
```
{
  _nitro: { routeRules: {} },
  nitro: { errors: [] },
  session: null,
  user: null,
  matchedRoute: { path: '/api/any-api-route', handlers: { get: [Function] } },
  params: {}
}
```

I was trying to figure why `$fetch` wasn't including the cookie in the server request, and spent quite some time to find `useRequestFetch` as it is not documented yet (cf #19815). If you are using Lucia and you are having this `H3Error: No session or user ID found` problem, this is the solution!

**Comment by krthr** [comment]
I had the same problem as @blqke. Because I didn't find info about `useRequestFetch` in the docs I did this: 

```ts
export default defineNuxtRouteMiddleware(async (to, from) => {
  const authCookie = useCookie("auth").value;
  const user = useUser();

  const headers = new Headers();
  if (authCookie) {
    headers.set("Cookie", `auth=${authCookie}`);
  }

  const data = await $fetch("/api/auth/user", { headers });
  if (data?.user) {
    user.value = data.user;
  }
});
```

mentioning `useRequestFetch` in the docs could be very useful!

**Comment by vibonacci** [comment]
`useRequestFetch()` only includes the browser cookies on the first SSR request. After hydration, the cookies are missing.

**Comment by mohammadGh** [comment]
How to handle this issue more generally and what should be done when we don't use fetch directly?!

Assume that we are using an SDK function in `useAsyncData` and this function uses its own version of `fetch`. For example: 

```js
<script setup>
const { data: userProfile } = await useAsyncData('user-profile', () => $api.getCurrentUser())
</script>
```
Assuming that login has already been performed and the appropriate `cookie` exists, the above function works correctly on the client-side. However, it encounters problems during server-side rendering. As @Atinux sayed we need to forward the browser's cookies to the API routes when fetching the data (our performing any request) on SSR.

**In general**, any information present in the context of a page request, including `headers` and `cookies`, that triggers the SSR process should be propagated to every request made within that page using the fetch API.
This aligns with the developer's perspective since they perceive all fetch calls as originating from the browser, which automatically includes headers and cookies. 

Dear  @danielroe and @pi0 , I would appreciate your thoughts on this matter. Also, what are some current solutions for the example I mentioned? That is, when using an SDK and not having direct access to fetch, what should we do to avoid SSR issues?

**Comment by Degoya-Studio** [comment]
I can't seem to get the session on my server side. 

`<script setup lang="ts">
async function handleLogout() {
  await useFetch('/api/auth/logout', {method: 'GET'})
}
</script>`

`export default defineEventHandler(async (event) => {
    await requireUserSession(event);`
    
any idea of what I could be doing wrong here?

**Comment by moshetanzer** [comment]
Have spent ages on this too. What is strange is that even when declaring a custom fetch just wrapped around useFetch like this

```
export function useSecureFetch<T>(
  url: string | (() => string),
  options?: UseFetchOptions<T>
) {
  return useFetch(url, {
    ...options,
    $fetch
  })
}
```

you will get an error since context isn't yet available on server side, though shouldnt this work just as useFetch which does work?

(use case is useSecureFetch with a custom $fetch, however even with normal $fetch on server fetch it doesnt work)

@danielroe is this expected?

thanks again!

**Comment by danielroe** [comment]
@moshetanzer No, that won't pass on the request details. In fact, useFetch passes on the request details by default and you are overriding _not_ to do it.

**Comment by TutorFx** [comment]
> We can also introduce a new composable like `$fetchWithHeaders`

example:

```ts 
// app/composables/fetch.ts
import type { FetchOptions } from 'ofetch'

export function $fetchWithHeaders(options: FetchOptions) {
  const headers = useRequestHeaders()

  return $fetch.create({
    ...options,
    headers,
  })
}
```

**Comment by alex-eri** [comment]
Forwarding cookie using `useFetch` not working in Nuxt 4.3.1. 
works in 4.2.2

**Comment by danielroe** [comment]
@alex-eri please open an issue with a reproduction 🙏
